### PR TITLE
Remove prop injection in Show page components

### PR DESCRIPTION
**Pull request by fzaninotto**
## Problem

The main page controller mechanism relies on prop injection, which is neither intuitive or practical:

```jsx
const PostShow = props => (
   <Show {...props}>
       <SimpleShowLayout>
```

`SimpleShowLayout` receives some props, but it's not apparent which ones. It makes it harder to create a custom layout.

## Solution

Use contexts to *pull* data instead of pushing it in the main controllers. This leads to the ability to do:

```jsx
import { Show, SimpleShowLayout, TextField, DateField, WithRecord } from 'react-admin';
import StarIcon from '@mui/icons-material/Star';

const BookShow = () => (
    <Show>
        <Grid container spacing={2}>
            <Grid item xs={12} sm={8}>
                <SimpleShowLayout>
                    <TextField label="Title" source="title" />
                    <DateField label="Publication Date" source="published_at" />
                    <WithRecord label="Rating" render={record => <>
                        {[...Array(record.rating)].map((_, index) => <StarIcon key={index} />)}
                    </>} />
                </SimpleShowLayout>
            </Grid>
            <Grid item xs={12} sm={4}>
                <Typography>Details</Typography>
                <Stack spacing={1}>
                    <Labeled label="ISBN"><TextField source="isbn" /></Labeled>
                    <Labeled label="Last rating"><DateField source="last_rated_at" /></Labeled>
                </Stack>
            </Grid>
    </Show>
);
```

## Tasks

- [x] [BC Break] Remove prop injection in `<Resource>` for the show view (Breaking change until we upgrade to react-router v6 to support basename), so Show components must now pull what they need from contexts
- [x] [BC Break] Remove the prop injection in `<Show>` prop, so Show Layout and Show Actions components must now pull what they need from contexts
- [x] [BC Break] Move the responsibility to fetch route params from `Resource` to `useShowController` (via react-router's `useParams`)
- [x] [BC Break] Move `Card` from `<Show>` to `<SimpleShowLayout>` and `<TabbedShowLayout>` 
- [x] [BC Break] Remove the `<Show aside>` prop (no longer needed as it's trivial) 
- [x] [BC Break] Move loading handling from `<Show>` to children (like in `<List>`)
- [x] [BC Break] `<SimpleShowLayout>` and `<TabbedshowLayout>` no longer use the `addLabel` prop to decide when to add a label to a field
- [x] Update Show documentation (note: this went way further than I expected)
- [x] Check that everything works using stories & integration tests
- [x] Add upgrade doc
-  ~~Move Labeled decoration responsibility to the Fields instead of the Layouts, based on a Context~~ (will render two labels in Reference fields)

### CODE REVIEW

**Comment by fzaninotto** [comment]
Switching to RFR

**Review by djhi — CHANGES_REQUESTED** [review]
Awesome!

File: packages/ra-core/src/controller/show/useShowController.ts [hunk 1]
```diff
@@ -50,39 +44,43 @@ export interface ShowControllerProps<RecordType extends Record = Record> {
...
+    props: ShowControllerProps
+): ShowControllerResult<RecordType> => {
+    const { id: propsId, onFailure } = props;
     const resource = useResourceContext(props);
     const translate = useTranslate();
     const notify = useNotify();
     const redirect = useRedirect();
     const refresh = useRefresh();
     const version = useVersion();
+    const { id } = useParams<{ id?: string }>();
```
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-core/src/controller/show/useShowController.ts:72]: I don't see the use case for reusing that type, or for reusing the hook outside of the Show controller. Nice to have.

[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/controller/show/useShowController.ts:72]: Agreed

**Review by djhi — APPROVED** [review]

**Pull request merged**